x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1237,109 +1261,97 @@ boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible)
+                    return true;
+                }
+
+                return opSpecificIsUnsafe();
             }
 
-            Expression doSimplify() {
+            Expression apply() {
                 // force float point folding for FlP field
                 Literal bcl = operation.dataType().isRational()
                     ? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())
                     : bcLiteral;
 
-                if (operation instanceof BinaryComparisonInvertible == false) { // should be an assertion?
-                    throw new QlIllegalArgumentException("Unexpected [" + operation.symbol() + "] operation object");
-                }
-                Expression bcRightExpression = ((BinaryComparisonInvertible) operation).binaryComparisonInverse()
+                Expression bcRightExpression = ((BinaryComparisonInversible) operation).binaryComparisonInverse()
                     .create(bcl.source(), bcl, opRight);
-                bcRightExpression = safeMaybeFold(bcRightExpression);
-                return bcRightExpression == null ? comparison : comparison.replaceChildren(List.of(opLeft, bcRightExpression));
+                bcRightExpression = tryFolding(bcRightExpression);
+                return bcRightExpression != null
+                    ? opSpecificPostApply((BinaryComparison) comparison.replaceChildren(List.of(opLeft, bcRightExpression)))
+                    : comparison;
             }
 
-            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
-                ArithmeticOperation operation = (ArithmeticOperation) comparison.left();
-                // Use symbol comp: SQL operations aren't available in this package (as dependencies)
-                String opSymbol = operation.symbol();
-                // Modulo can't be simplified.
-                if (MOD.symbol().equals(opSymbol)) {
-                    return comparison;
-                }
-                SimplifyOperation simplify = null;
-                if (isMulOrDiv(opSymbol)) {
-                    simplify = new SimplifyMulDiv(comparison);
-                } else if (ADD.symbol().equals(opSymbol) || SUB.symbol().equals(opSymbol)) {
-                    simplify = new SimplifyAddSub(comparison);
-                }
-
-                return (simplify == null || simplify.cannotSimplify(typesCompatible)) ? comparison : simplify.doSimplify();
-            }
+            // operation-specific operations:
+            //  - fast-tracking of simplification unsafety
+            abstract boolean opSpecificIsUnsafe();
 
-            static boolean isMulOrDiv(String opSymbol) {
-                return MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol);
+            //  - post adjustments
+            Expression opSpecificPostApply(BinaryComparison binaryComparison) {
```

Comment:
postApply or postProcess. opSpecific is somewhat meaningless.

Reply:
I went for `postProcess`.